src/action.c: Convert action->arg to a list of action_arg

Consolatis opened the pull request:
In preperation for Actions that require multiple arguments

Consolatis commented on include/action.h:
```diff
@@ -2,19 +2,48 @@
+void action_add_int_arg(struct action *action, char *name, int value);
+void action_add_double_arg(struct action *action, char *name, double value);
+void action_add_bool_arg(struct action *action, char *name, bool value);
+struct wlr_box *action_add_box_arg(struct action *action, char *name);
```
Only `action_add_str_arg()` is used in this PR, the others are examples of how this action_arg framework could be used

Consolatis commented on include/private/action.h:
```diff
@@ -0,0 +1,41 @@
+struct action_arg_box {
+	struct action_arg base;
+	struct wlr_box value;
+};
```
Only `struct action_arg_str` is used in this PR, the others are examples of how this action_arg framework could be used in the future

Consolatis commented on src/action.c:
```diff
@@ -56,6 +60,52 @@ const char *action_names[] = {
+{
+	assert(arg->type == LAB_ACTION_ARG_BOX);
+	return &((struct action_arg_box *)arg)->value;
+}
```
Only `action_str_from_arg()` is used in this PR, the others are examples of how this action_arg framework could be used in the future

Consolatis commented on src/action.c:
```diff
@@ -267,9 +335,70 @@ actions_run(struct view *activator, struct server *server,
+	}
+	wl_list_insert(action->args.prev, &arg->base.link);
+	return &arg->value;
+}
```
Only `action_add_str_arg()` is used in this PR, the others are examples of how this action_arg framework could be used in the future

johanmalm commented on include/action.h:
```diff
@@ -2,19 +2,48 @@
 #define __LABWC_ACTION_H
 
-struct server;
+/* Forward declare arguments */
```
Consider "/* Forward declaration of opaque structs for pointer arguments */"
Although I think I prefer leaving it out because we should try to do that wherever we can.

Consolatis replied:
Will remove it completely then

johanmalm commented on include/action.h:
```diff
@@ -2,19 +2,48 @@
+	                       * struct mousebinding.actions,
+	                       * struct menuitem.actions */
+
+	uint32_t type;        /* enum action_type */
```
Wouldn't it be better to declare the enum in this header-file and just make this `enum action_type type`.
I'm thinking that other translation units will want to do things like `if (action->type == LAB_ACTION_WHATEVER)`

Consolatis replied:
I don't know if we actually need that kind of introspection outside of `action.c`.
The idea was that nobody else should care about the internals of `action.c`. The outside API is either string based (`action_create()`) or `struct action *` based so the enum isn't ever used anywhere other than in `action.c` itself.

The action names are internally translated into the enum in `action_create()` and `actions_run()` gets a *wl_list of action structs as argument.
Maybe we could start using the same notation as wlroots in just putting  a `/* private */` on top of struct members like this?

johanmalm commented on include/action.h:
```diff
@@ -2,19 +2,48 @@
+	struct wl_list link;  /* struct action.args */
 
+	/**
+	 * FIXME: Do we want to replace `name` by a plain uint32_t and have
```
Don't mind. action_run() isn't a function that gets called that often, so the tiny overhead of a strcmp()s probably pales into insignificance :smile:

johanmalm commented on include/action.h:
```diff
@@ -2,19 +2,48 @@
+
+
+struct action *action_create(const char *action_name);
+void action_add_str_arg(struct action *action, char *name, const char *value);
```
Would be more consistent to name these functions `action_add_arg_XX` or even `action_arg_add_XX`

Consolatis replied:
will change to `action_arg_add_XX()`

johanmalm commented on include/action.h:
```diff
@@ -2,19 +2,48 @@
+void action_list_free(struct wl_list *action_list);
 
-#endif
+#endif // __LABWC_ACTION_H
```
Prefer `/* __LABWC_ACTION_H */` style comments

Consolatis replied:
woops :)

johanmalm commented on include/config/keybind.h:
```diff
@@ -11,8 +11,8 @@ struct keybind {
 	size_t keysyms_len;
-	struct wl_list actions;
-	struct wl_list link;
+	struct wl_list actions;  /* action::link */
```
We ought to stick with either `::` or `.`
I think I saw a change in wlroots recently moving from `::` to `.` which I guess more C like. I don't really mind which, but consistency is good. (The `.` was used in action.h above)

Consolatis replied:
woops. Saw the wlroots change as well but for some reason still used `::` here.

johanmalm commented on include/config/keybind.h:
```diff
@@ -11,8 +11,8 @@ struct keybind {
 	size_t keysyms_len;
-	struct wl_list actions;
-	struct wl_list link;
+	struct wl_list actions;  /* action::link */
```
I think that for `wl_list actions` is obvious what it's a list of.... a list of actions.
So I think the only `struct wl_list` entries where it's useful to have a comment is the `struct wl_list link` because it's not obvious what list is using that link... so the line below `/* rcxml::keybinds */` is where it's useful.
Do you agree?

Consolatis replied:
If i think about being new to the codebase and seeing `wl_list actions` I'd also assume there are some actions in that list. But I wouldn't necessarily know that there is a `struct action {` I could grep for in `include`.
As comments are free I'd prefer to have `/* struct action.link */` in there.

johanmalm replied:
Yes, good call. Simple thing to help readability.

johanmalm commented on src/action.c:
```diff
@@ -200,11 +266,15 @@ actions_run(struct view *activator, struct server *server,
 			break;
 		case ACTION_TYPE_RECONFIGURE:
-			/* Should be changed to signal() */
+			/* Could likely be changed to kill(0, SIGHUP) */
```
Oh nice :smile:

johanmalm commented on src/action.c:
```diff
@@ -56,6 +60,52 @@ const char *action_names[] = {
+}
+
+static struct action_arg *
+action_get_first_arg(struct action *action)
```
In terms of the general design, what do you think of the following???

```
struct action {
	char *name;
	struct wl_list arguments;

	struct wl_list link; /* keybind.actions, etc */
}

struct action_arg {
	enum action_arg_type type;
	char *key;	/* e.g. "width", "direction" */
	union {
		char *value;
		int value;
		etc...
	}
	struct wl_list link; /* action.arguments */
}
```

It's possibly a bit simpler and behave more like a dictionary (although implemented as a `wl_list`).

The argument key gives the option of a helper in action.c to do things like `width = arg(action_arg_list, "width")`
It avoids functions like `get_first_argument`

Just a thought :smile:

Consolatis replied:
I like this better than the inheritance thing. Need to think a bit about this.

Re `width = arg(action_arg_list, "width")`:
How would the function signature for `arg()` look like in this case? At some point it would need to cast the result to one of the union members?

Consolatis replied:
Re union vs inheritance + casting, gave the union a try and ended up with this:
```c
struct action_arg {
        const char *key;      /* May be NULL if there is just one arg */
        struct wl_list link;  /* struct action.args */

        /* Private */
        uint32_t type;        /* enum action_arg_type */
        union {
                char *val_str;
                int val_int;
                double val_double;
                bool val_bool;
                struct wlr_box val_box;
        };
};
```
It makes the internal `action_x_from_arg()` function a bit easier to read but IMHO also
- exposes internal members to the outside
- make the struct look ugly

Are there technical reasons to prefer one over the other?

I can only think of a positive for the inheritance + casting approach:
- it only needs as much additional storage as required be the relevant argument instead of the max size of any argument

johanmalm reviewed (commented):
I like it in principle. Mostly minor comments, but one change in design proposal (just structuring the structs slightly differently)

Consolatis commented:
I think I addressed everything, if I missed something or there is something else which should be changed / improved please say so. Also removed the not (yet?) required variants for `int`, `bool`, `double` and `wlr_box` for now.
